### PR TITLE
Added an optional user modifiable default sketch file when creating a new project.

### PR DESCRIPTION
**Pull request by nbourre**
### Motivation
Sometimes programmers use the same boilerplate code for their project. This PR allows the creation of a default sketch.

### Change description
If a `default/default.ino` file is present in the Arduino Sketch folder, the app will read the content of the `default.ino` file to create the new project. If the file is non-existent, the default hard-coded sketch is used.

### Other information
I created a `loadDefault()` method in the `sketches-service-impl.ts` file.

### Reviewer checklist

* [x] PR addresses a single concern.
* [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [x] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by nmzaheer** [comment]
This needs to be documented somewhere for user discoverability.

As a further change, we could add a settings parameter that points to a path of the template sketch in a folder marked for templates. This would make managing of templates easier.

**Comment by per1234** [comment]
> This needs to be documented

Excellent point @nmzaheer. I would suggest the "Advanced Usage" document as the appropriate place for such information:

https://github.com/arduino/arduino-ide/blob/main/docs/advanced-usage.md

However, that would not be necessary if your other suggestion was implemented:

>add a settings parameter

In this case, I think the existence of the setting and the accompanying description would be sufficient to make the feature discoverable to the user.

**Comment by nmzaheer** [comment]
> However, that would not be necessary if your other suggestion was implemented:

So are you suggesting that this change should not be merged and rather we work a bit more to add it to the Settings ?

**Comment by nbourre** [comment]
Could it be an iterative update. Such as in the first iteration, it is not in the UI but only in the documentation and in a second iteration, the UI is available in an advanced settings?

**Comment by kittaakos** [comment]
> Could it be an iterative update.

No question we need them together.

> I don't feel comfortable making change to the UI.

You do not have to. Checking your changeset, I am positive you can do it. Here are a few pointers:

1. If you want to add support for a new setting, you need to update the "schema", and the framework will generate the rest. Nice, isn't it? You have to change ` /arduino-ide-extension/src/browser/arduino-preferences.ts` here:

```diff
diff --git a/arduino-ide-extension/src/browser/arduino-preferences.ts b/arduino-ide-extension/src/browser/arduino-preferences.ts
index 5fef5907..9e9b3f9d 100644
--- a/arduino-ide-extension/src/browser/arduino-preferences.ts
+++ b/arduino-ide-extension/src/browser/arduino-preferences.ts
@@ -241,6 +241,14 @@ export const ArduinoConfigSchema: PreferenceSchema = {
       ),
       default: false,
     },
+    'arduino.sketch.inoBlueprint': {
+      type: 'string',
+      description: nls.localize(
+        'arduino/preferences/sketch/inoBlueprint',
+        'Absolute filesystem path to the default `.ino` blueprint file. If specified, the content of the blueprint file will be used for every new sketch created by the IDE. The sketches will be generated with the default Arduino content if not specified. Unaccessible blueprint files are ignored. A restart of the IDE is needed for this setting to take effect.'
+      ),
+      default: undefined,
+    },
     'arduino.checkForUpdates': {
       type: 'boolean',
       description: nls.localize(
@@ -278,6 +286,7 @@ export interface ArduinoConfiguration {
   'arduino.auth.registerUri': string;
   'arduino.survey.notification': boolean;
   'arduino.cli.daemon.debug': boolean;
+  'arduino.sketch.inoBlueprint': boolean;
   'arduino.checkForUpdates': boolean;
 }
```

2. Next task is to read and use this value from the settings. Usually, these things happen in the browser window through a proper service, but we need to use a hack here. No worries, we did this once. Here is an example:

https://github.com/arduino/arduino-ide/blob/99b10942bb944d3916819aac41a5ede94c93fead/arduino-ide-extension/src/node/arduino-daemon-impl.ts#L174-L194

3. Make sure you read the content of the blueprint file once and save it as a field to the `SketchesService`. Currently, you read it every time you invoke `loadDefault`. It's not good. State in the description of the new setting that IDE2 needs a restart. It's OK for now.

4. I am OK ignoring all errors when reading the blueprint file (`fs.readFile`), but if the `code` is not `ENOENT` log it to the console at least. See the example above.

5. When you're done with the development, run `yarn i18n:generate` from the repo's root to update the translation files with the description of the new setting.

6. I recommend you install [`dbaeumer.vscode-eslint`](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) VS Code extension if you're using Code. It will clean up all styling issues, such as `let raw = "";` should be `let raw = '';`

Keep up the great work, and let me know if you have any questions.

**Comment by kittaakos** [comment]
We should probably consider the default content for the remote sketches:

https://github.com/arduino/arduino-ide/blob/93291b68119cf46974491a727cb342ec53fcb3d4/arduino-ide-extension/src/browser/create/create-api.ts#L521-L533

**Comment by kittaakos** [comment]
@nbourre, do you plan to make the requested changes? We want to release a new patch version of IDE2 soon and include this feature.

**Comment by nmzaheer** [comment]
@kittaakos I would like to take a look at this. How do I contribute without creating a new PR? Should I open a PR in @nbourre repo and if that is merged then it would be reflected in this PR ?

On a side note, I think we need to start tagging issues to milestones i.e. 2.01, 2.02, etc. for contributors to know what i being prioritized by the development team.

**Comment by nbourre** [comment]
@kittaakos What is the ETA for the next release? I'm quite tight in my time today. I might be able to do it this weekend.

@nmzaheer I will try to work on it this weekend, but if you have time and are fluent with this framework, feel free to do a PR on my repo.

**Comment by kittaakos** [comment]
> @kittaakos What is the ETA for the next release?

Thank you for the quick reply. I plan to merge everything by latest next Tuesday (25.10).

If your schedule is tight, I am happy to add support for the `settings.json` as a separate commit and keep you as the original author of the PR. I am also glad to answer all your questions if you plan to solve the task alone.

You're very close to the final solution, and I think you can do it. 💪 But no pressure, of course.

**Comment by nbourre** [comment]
Hi guys,

Link to my last commit : https://github.com/arduino/arduino-ide/commit/6444edc0c6609dff8aace4f1c6dda339febb8008

I'm having trouble to make the modification work.

By @kittaakos, here are the task and what I've done so far :

- [x] Create a setting in the UI
- [x] Read the value from the settings (the hack)
- [x] Use a field for the blueprint content.
- [x] Catch error with `ENOENT`
- [ ] Generate i18n
- [ ] ES linter
- [ ] Documentation

But I don't see the settings in the preferences. What am I missing and the default file is not loaded anymore. 😢 

The commands I've done :

```cmd
yarn
yarn rebuild:browser
yarn rebuild:electron
yarn start
```

Do you guys have a cue to improve the build speed? Building the apps takes an eternity compared to C-style apps.

Thanks

**Comment by nbourre** [comment]
Ok, I got it. The default folder is going to be in the advanced `Preferences` not the one accessible by the menu.

**Comment by nbourre** [comment]
@kittaakos I think I have accomplished all the requirements for the next release.

File: arduino-ide-extension/src/browser/arduino-preferences.ts [hunk 1]
```diff
@@ -249,6 +249,14 @@ export const ArduinoConfigSchema: PreferenceSchema = {
...
       default: true,
     },
+    'arduino.sketch.inoBlueprint': {
+      type: 'string',
+      markdownDescription: nls.localize(
+      'arduino/preferences/sketch/inoBlueprint',
+      'Absolute filesystem path to the default `.ino` blueprint file. If specified, the content of the blueprint file will be used for every new sketch created by the IDE. The sketches will be generated with the default Arduino content if not specified. Unaccessible blueprint files are ignored. **A restart of the IDE is needed** for this setting to take effect.'
+      ),      
+      default: undefined,
+      },
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/arduino-preferences.ts:259]: This is incorrectly formatted.

File: arduino-ide-extension/src/node/sketches-service-impl.ts [hunk 2]
```diff
@@ -637,6 +628,62 @@ void loop() {
...
+    let result: string;
+    result = `void setup() {
+  // put your setup code here, to run once:
+
+}
+      
+void loop() {
+  // put your main code here, to run repeatedly:
+  
+}`;
```
Comment (kittaakos) [inline arduino-ide-extension/src/node/sketches-service-impl.ts:656]: There are undesired whitespaces in the template.

File: arduino-ide-extension/src/node/sketches-service-impl.ts [hunk 3]
```diff
@@ -637,6 +628,62 @@ void loop() {
...
+        encoding: 'utf8',
+      });
+      const json = this.tryParse(raw);
+      if (json) {
+        const value = json['arduino.sketch.inoBlueprint'];
+
+        const filename =
+          typeof value === 'string' && !!value
+            ? value
+            : `${root}/default/default.ino`;
```
Comment (kittaakos) [inline arduino-ide-extension/src/node/sketches-service-impl.ts:669]: This is something IDE2 won't support. It's either given by the user, or the default is used from the code.

**Review by kittaakos — CHANGES_REQUESTED** [review]
> Ok, I got it. The default folder is going to be in the advanced `Preferences` not the one accessible by the menu.

There is no default folder. Users either give an absolute path to the default `.ino` content file or the default will be used from the code. But IDE2 won't try to pick up any files from `#directories.user/default/default.ino`.

Please check your template, you have added undesired whitespaces. See [here](https://github.com/arduino/arduino-ide/pull/1559/files#diff-02510705818960c89dfce2311fd75354583e15e84a49fe8928a86bfa9877bcc7R649-R656).

I simplified the code. Here is my proposal: https://github.com/nbourre/arduino-ide/compare/main...kittaakos:arduino-ide:default-ino-review

Feel free to use it.

File: arduino-ide-extension/src/browser/arduino-preferences.ts (hunk 1 above)
Comment (nbourre) [reply arduino-ide-extension/src/browser/arduino-preferences.ts:259]: In what way? The `markdownDescription` with the markdown content?

**Comment by nbourre** [comment]
@kittaakos, except for the description which I'm not sure what's wrong. I've applied the requested changes.

**Comment by kittaakos** [comment]
> except for the description which I'm not sure what's wrong.

The indentation of the code was incorrect, but you have fixed it. Thank you!

All looks good except the default template.

If the new `arduino.sketch.inoBlueprint` setting is not defined or pointing to an inaccessible file, the default `.ino` content looks like this using the build from this PR:

```cpp
void setup() {
  // put your setup code here, to run once:
}
void loop() {
  // put your main code here, to run repeatedly:
}
```

Expected:
```cpp
void setup() {
  // put your setup code here, to run once:

}

void loop() {
  // put your main code here, to run repeatedly:

}

```

 - missing empty line after comment inside the method,
 - missing empty line between methods, and
 - missing newline at the end of the file.


> Here is my proposal: [nbourre/arduino-ide@main...kittaakos:arduino-ide:default-ino-review](https://github.com/nbourre/arduino-ide/compare/main...kittaakos:arduino-ide:default-ino-review)

Please reference my proposed template [here](https://github.com/nbourre/arduino-ide/compare/main...kittaakos:arduino-ide:default-ino-review#diff-02510705818960c89dfce2311fd75354583e15e84a49fe8928a86bfa9877bcc7R38-R46). If you fix the `.ino` content, this PR will be ready for merge.

**Review by kittaakos — APPROVED** [review]
It's working great. Thank you so much for your contribution, @nbourre 👍

**Comment by kittaakos** [comment]
I have also verified on Windows from the build. ✨

**Comment by UKHeliBob** [comment]
This seems to have been made more complicated that needed by adding a new setting.  It is too late now, but what was the problem with the original proposal to use a sketch named default,ino if it existed rather than adding a new setting ?

Come to that you could simply have used the same mechanism as the classic IDE and used the BareMinimum example as the new sketch template

Anyone who felt the need for a different default sketch, as I did, would simply edit the BareMinimum example to suit themselves

**Comment by per1234** [comment]
Hi @UKHeliBob 

>This seems to have been made more complicated that needed by adding a new setting.

I don't see any problem with the feature being slightly complex to set up. This is something only a small portion of advanced users will do, and something they will do only once.

The important thing is to provide the capability, and to provide that capability in a manner that can meet all user's needs, not in streamlining the feature as would be more of a consideration for some operation that is being done repeatedly or by beginners.

>what was the problem with the original proposal to use a sketch named default,ino if it existed

Because some users might create a sketch of this name without any intention of it being adopted by the IDE as the base sketch.

Because some users might prefer to store the sketch in some other location or under a different name.

>you could simply have used the same mechanism as the classic IDE and used the BareMinimum example as the new sketch template

The reimplementation of each feature in Arduino IDE 2.x is a chance to reconsider the old one and improve on it where areas for improvement are identified. This was one with significant areas for improvement, and those who collaborated here on the feature took the time and effort to make it better!

With the old system, your custom new sketch was lost at every IDE update.

With some operating systems, packages, security configurations, it is impossible to edit files inside the Arduino IDE installation package.

Some users may have different requirements for the new sketch than they do for the sketch accessed via **File > Examples > 01.Basics > BareMinimum** example sketch.

**Comment by nmzaheer** [comment]
The whole point of IDE 2 is that it will have breaking changes compared to IDE 1. However, the goal is not to alienate the entire userbase by creating a user experience which is vastly different from IDE 1 since that is the targeted audience.

As pointed out by @per1234 , the Bare minimum sketch gets overwritten with every update of Arduino IDE 2 which does not ensure a good UX since the user will have to keep changing contents with every update. And I am assuming till features and 🐛stabilize, the team would be looking at having release every month or so.

So, I feel having a setting to a default sketch would be a good UX rather than editing the BareMinimum sketch.

On a side note, maybe the team could also discuss having a menu item under **File** with **New Sketch from Template** as an option. However, we will have to set up some sort of data collection to understand how many use that option and if it deserves such high visibility under the **File** menu

**Pull request merged**